contracts/AddressBook.sol

```diff
@@ -3,10 +3,233 @@
+    /// @dev oracle key
+    bytes32 private constant ORACLE = "ORACLE";
+    /// @dev weth token key
+    bytes32 private constant WETH = "WETH";
+
+    /// @dev a mapping between key and address
+    mapping(bytes32 => address) private addresses;
+
+    /// @notice event emitted when a new proxy get created
+    event ProxyCreated(bytes32 id, address proxy);
+    /// @notice event emitted when a new address get added
+    event AddressAdded(bytes32 id, address add);
+
+    /**
+     * @notice return an address for specific key
+     * @param _key key address
+     * @return address
+     */
+    function getAddress(bytes32 _key) public view returns (address) {
+        return addresses[_key];
+    }
+
+    /**
+     * @notice return otoken implementation address
+     * @return otoken implementation address
+     */
+    function getOtokenImpl() external view returns (address) {
+        return getAddress(OTOKEN_IMPL);
+    }
+
+    /**
+     * @notice return otoken factory address
+     * @return otoken factory address
+     */
+    function getOtokenFactory() external view returns (address) {
+        return getAddress(OTOKEN_FACTORY);
+    }
+
+    /**
+     * @notice return whitelist address
+     * @return whitelist address
+     */
+    function getWhitelist() external view returns (address) {
+        return getAddress(WHITELIST);
+    }
+
+    /**
+     * @notice return controller address
+     * @return controller address
+     */
+    function getController() external view returns (address) {
+        return getAddress(CONTROLLER);
+    }
+
+    /**
+     * @notice return pool address
+     * @return pool address
+     */
+    function getMarginPool() external view returns (address) {
+        return getAddress(MARGIN_POOL);
+    }
+
+    /**
+     * @notice return margin calculator address
+     * @return margin calculator address
+     */
+    function getMarginCalculator() external view returns (address) {
+        return getAddress(MARGIN_CALCULATOR);
+    }
+
+    /**
+     * @notice return liquidation manager address
+     * @return liquidation manager address
+     */
+    function getLiquidationManager() external view returns (address) {
+        return getAddress(LIQUIDATION_MANAGER);
+    }
+
+    /**
+     * @notice return oracle address
+     * @return oracle address
+     */
+    function getOracle() external view returns (address) {
+        return getAddress(ORACLE);
+    }
+
+    /**
+     * @notice return WETH token
+     * @return WETH address
+     */
+    function getWeth() external view returns (address) {
+        return getAddress(WETH);
+    }
+
+    /**
+     * @notice set a specific address for a specific key
+     * @dev can only be called by addressbook owner
+     * @param _key key
+     * @param _address address
+     */
+    function setAddress(bytes32 _key, address _address) public onlyOwner {
+        addresses[_key] = _address;
+
+        emit AddressAdded(_key, _address);
+    }
```

Comment:
We should put this after external function.

- Recommended order of functions in a contract: 
    - constructor
    - fallback
    - external
    - public
    - internal
    - private